### PR TITLE
Graph request serializer

### PR DESCRIPTION
**Pull request by joesus**
## Checklist

- [x] I've read the [Contributing Guidelines](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I've completed the [Contributor License Agreement](https://developers.facebook.com/opensource/cla)
- [x] I've ensured that all existing tests pass and added tests (when/where necessary)
- [x] I've ensured that my code lints properly: (`swiftlint && swiftlint autocorrect --format`)
- [x] I've updated the documentation (when/where necessary) and [Changelog](CHANGELOG.md) (when/where necessary)
- [x] I've added the proper label to this pull request (e.g. `bug` for bug fixes)

## Pull Request Details

**Background:**
In the original implementation an `FBSDKGraphRequest` had a number of serialization methods for taking a `String` representation of a `URL`, adding escaped query parameters from the `GraphRequest` and `Settings`, and returning a url-escaped (serialized) `String`. 

This pr is my attempt to update this approach to take a `URL` instead of a `String` and to use `URLQueryItem`s built from the `GraphRequest`'s parameters instead of a `String` built from the `GraphRequest`'s parameters. 

So now instead of taking a `String` and returning a `String`, this takes a `URL` and returns a `URL` with proper `URLQueryItem`s. Therefore I'm not sure it's accurate to call it a Serializer but would welcome suggestions on what else to call it. It may be useful to call it Serializer and have a task to update the naming later since the fewer name changes the better when introducing additional converted types.

## Test Plan

Test Plan: All tests should pass. No new linting errors. Coverage should remain near 100%

### CODE REVIEW

**Comment by joesus** [comment]
Closed and reopened to trigger build

File: Sources/Core/GraphRequestSerializer.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   Serializes a graph request and a url
+
+   - Parameters:
+     - url: The url to modify based on the graph request
+     - graphRequest: The request used to provide information to a url
+     - forBatch: Whether or not a graph request is intended to be batched with other requests
+   */
+  func serialize(with url: URL, graphRequest: GraphRequest, forBatch: Bool = false) throws -> URL {
+    if graphRequest.httpMethod == .post, !forBatch {
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]: Guard requires return. Always prefer a guard statement when returning early.

```suggestion
    guard graphRequest.httpMethod != .post || forBatch else {
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      return url
+    }
+
+    if graphRequest.hasAttachments,
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
    guard !graphRequest.hasAttachments
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      graphRequest.httpMethod == .get {
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
      || graphRequest.httpMethod != .get else {
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      logger.log(message: "Can not use GET to upload a file")
+      throw GraphRequestSerializationError.getWithAttachments
+    }
+
+    let requestQueryItems = GraphRequestQueryItemBuilder.build(from: graphRequest.parameters)
+    let processedQueryItems = preProcess(requestQueryItems)
+
+    guard var components = URLComponents(url: url, resolvingAgainstBaseURL: true) else {
+      throw GraphRequestSerializationError.malformedURL
+    }
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
    }

```

File: Sources/Core/GraphRequestSerializer.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    // Could not find other examples where this can fail. Since we are not modifying the path
+    // as part of this helper it is safe to assume this will always succeed.
+    return components.url ?? url
+  }
+
+  /**
+   Returns an updated list of parameters that includes information about debug levels
+
+    - Parameters:
+      - parameters: a list of `URLQueryItem`s to use in constructing a new list of `URLQueryItem`
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
    - Parameter parameters: a list of `URLQueryItem`s to use in constructing a new list of `URLQueryItem`
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    // Getting a url from the components can fail if we modify the components path parameter
+    // Could not find other examples where this can fail. Since we are not modifying the path
+    // as part of this helper it is safe to assume this will always succeed.
+    return components.url ?? url
+  }
+
+  /**
+   Returns an updated list of parameters that includes information about debug levels
+
+    - Parameters:
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      - parameters: a list of `URLQueryItem`s to use in constructing a new list of `URLQueryItem`
+                    that includes information about debug levels
+   */
+  func preProcess(_ parameters: [URLQueryItem]) -> [URLQueryItem] {
+    switch settings.graphApiDebugParameter {
+    case .none:
+      return parameters
+
+    case .info, .warning:
+      let queryItem = URLQueryItem(
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]: Avoid creation of variable

```suggestion
      return parameters + [
        URLQueryItem(
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        name: Keys.debug.rawValue,
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
          name: Keys.debug.rawValue,
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        value: settings.graphApiDebugParameter.rawValue
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
          value: settings.graphApiDebugParameter.rawValue
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      )
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
        )
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+      return parameters + [queryItem]
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]:
```suggestion
      ]
```

File: Sources/Core/GraphRequestSerializer.swift (hunk 2 above)
Comment (joesus) [reply Sources/Core/GraphRequestSerializer.swift]: 👍

File: Sources/Core/GraphRequestSerializer.swift [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  /**
+   Serializes a graph request and a url
+
+   - Parameter url: The url to modify based on the graph request
+   - Parameter graphRequest: The request used to provide information to a url
+   - Parameter forBatch: Whether or not a graph request is intended to be batched with other requests
+   */
+  func serialize(with url: URL, graphRequest: GraphRequest, forBatch: Bool = false) throws -> URL {
+    guard graphRequest.httpMethod != .post
+      || forBatch
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]: Suggested style difference

```suggestion
      || forBatch else {
```

File: Sources/Core/GraphRequestSerializer.swift [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+      else {
```
Comment (codytwinton) [inline Sources/Core/GraphRequestSerializer.swift]: Suggested style difference

```suggestion
```

**Pull request merged**